doc/decisions/transformations.md

```diff
@@ -0,0 +1,102 @@
+# Transformations
+
+## Problem
+
+Certain plugins, e.g. `rename`, perform extensive transformations on keys and keysets within Elektra.
+Those transformations include, but are not limited to:
+
+- Changing the names of the keys inbetween `kdbGet` and `kdbSet` operations
+- Changing values back and forth for normalization, e.g. `true` -> `1`, `1` -> `true`
+- Adding and removing metadata (e.g. the `spec` plugin)
+
+While these features are useful, they do create feature-interaction problems.
+More specifically, problems have been observed in conjunction with the following (overlapping) types of plugins:
+
+- notification plugins
+- plugins that do change tracking
+
+The problem, in general, can be described as: Which phase of the KDB should be used for notifications/change tracking?
+
+### Observed problems with changing key names
+
+For example, the [`rename` plugin](../../src/plugins/rename/README.md) supports different transformations for `get` and `set`.
+The plugin is executed in the post-storage phase for the `get` operation and the pre-storage phase for the `set` operation.
+We can instruct it to convert the keynames to lowercase in the `get` operation, and to UPPERCASE in the `set` operation.
+This results in the keys being in UPPERCASE in the configuration files, but they are presented in lowercase to other plugins and applications using the Elektra API.
+
+For example, here is a configuration file with a hypothetical format:
+
+```
+/DISPLAY/BRIGHTNESS = 100
+``` 
+
+As can be seen, the keys are in UPPERCASE within the configuration file.
+In Elektra keys are case-sensitive. 
+As operations on keysets such as `ksLookup` operate after the post-storage phase, `kdb get /DISPLAY/BRIGHTNESS` will fail.
+For Elektra, the key `/DISPLAY/BRIGHTNESS` does not exist, as the `rename` plugin transformed this into the lowercase `/display/brightness`.
+
+This leads to problems with the notification plugins.
+As notification plugins are executed after the post-storage phase of the `set` operation, they will receive a keyset with the already transformed keys.
```

Comment:
Don't know how to phrase this properly, but the issue is not "when" the plugins are executed, but which data they key. Executing them after `poststorage` is fine. You could still give them the "elektra-visible" data. You have to return that data from `kdbSet` after all, so the data does exist.

Reply:
Yes, I thougt about this too. Basically, we do a deep-dup of the data before we hand it off to the backends? And use this data to return from `kdbSet` and for the notification plugins.

Reply:
I'm not sure what you mean. [What currently should happen is](https://github.com/ElektraInitiative/libelektra/blob/master/doc/dev/kdb-operations.md#set-operation):

- The `KeySet` passed to `kdbSet` is not modified (expect for adding of metadata by `spec`)
- Backends receive parts of a deep-dup copy of this `KeySet`
- The deep-dup copy is discarded at the end